File: src/main/java/com/aric3435/temporaltrials/network/LoopStatePayload.java
package com.aric3435.temporaltrials.network;

import com.aric3435.temporaltrials.TemporalTrialsMod;
import net.minecraft.network.PacketByteBuf;
import net.minecraft.network.codec.PacketCodec;
import net.minecraft.network.packet.CustomPayload;
import net.minecraft.util.Identifier;

/**
 * Typed CustomPayload used by the Fabric play-payload API in these mappings.
 * Provides ID and CODEC so ClientPlayNetworking.registerGlobalReceiver can use it.
 */
public final class LoopStatePayload implements CustomPayload {

    public static final Id<LoopStatePayload> ID =
            new Id<>(Identifier.of(TemporalTrialsMod.MOD_ID, "loop_state"));

    public static final PacketCodec<PacketByteBuf, LoopStatePayload> CODEC =
            PacketCodec.of(LoopStatePayload::write, LoopStatePayload::read);

    private final boolean active;
    private final int day;
    private final long remainingTicks;
    private final boolean showIntro;

    public LoopStatePayload(boolean active, int day, long remainingTicks, boolean showIntro) {
        this.active = active;
        this.day = day;
        this.remainingTicks = remainingTicks;
        this.showIntro = showIntro;
    }

    public boolean isActive() {
        return active;
    }

    public int getDay() {
        return day;
    }

    public long getRemainingTicks() {
        return remainingTicks;
    }

    public boolean shouldShowIntro() {
        return showIntro;
    }

    @Override
    public Id<? extends CustomPayload> getId() {
        return ID;
    }

    private static void write(LoopStatePayload payload, PacketByteBuf buf) {
        buf.writeBoolean(payload.active);
        buf.writeVarInt(payload.day);
        buf.writeVarLong(payload.remainingTicks);
        buf.writeBoolean(payload.showIntro);
    }

    private static LoopStatePayload read(PacketByteBuf buf) {
        boolean active = buf.readBoolean();
        int day = buf.readVarInt();
        long remaining = buf.readVarLong();
        boolean showIntro = buf.readBoolean();
        return new LoopStatePayload(active, day, remaining, showIntro);
    }
}

File: src/main/kotlin/com/aric3435/temporaltrials/LoopStatePayload.kt
package com.aric3435.temporaltrials

import net.minecraft.network.PacketByteBuf
import net.minecraft.network.codec.PacketCodec
import net.minecraft.network.packet.CustomPayload

class LoopStatePayload(
  val isActive: Boolean,
  val startTime: Long,
  val showScreen: Boolean
) : CustomPayload {
  override fun getId() = ID

  companion object {
    // ↓ use the shared constant, NOT a private field in your mod class
    val ID = CustomPayload.Id<LoopStatePayload>(Constants.LOOP_STATE_ID)

    val CODEC: PacketCodec<PacketByteBuf, LoopStatePayload> = PacketCodec.of(
      { pkt, buf ->
        buf.writeBoolean(pkt.isActive)
           .writeLong(pkt.startTime)
           .writeBoolean(pkt.showScreen)
      },
      { buf ->
        LoopStatePayload(
          buf.readBoolean(),
          buf.readLong(),
          buf.readBoolean()
        )
      }
    )
  }
}